### PR TITLE
Add option to disable IMDS v1 fallback when token is not returned

### PR DESCRIPTION
**Pull request by cenedhryn**
## Motivation and Context
Currently, InstanceProfileCredentialsProvider makes credentials call to the EC2 Instance Metadata Service (IMDS) by first attempting to fetch a token which is then attached to the credentials call, to provide better security. If token fetching fails due to some errors (403, 404, 405) or timeouts, the provider makes an unsecure call without the token header. 

For some use cases, this fallback behavior with unsecure, IMDS v1 style calls without a token is forbidden. To facilitate disabling this fallback, this PR adds a configuration option. 
 
Note that currently, it's only IMDS credential calls that honor the config option. Other calls to IMDS for region etc are not affected. 

## Modifications
- Adds setting to disable IMDS v1 fallback through 
  - Environment variables (`AWS_EC2_METADATA_V1_DISABLED`)
  - System property (`aws.disableEc2MetadataV1`) 
  - Config file (`ec2_metadata_v1_disabled`). 
- Updates creating an InstanceProfileCredentialsProvider in ProfileCredentialsProvider to forward profile information to the InstanceProfileCredentialsProvider

## Testing
Unit tested

Rewrote existing test to JUnit5

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [x] My change requires a change to the Javadoc documentation
- [x] I have updated the Javadoc documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/auth/src/main/java/software/amazon/awssdk/auth/credentials/InstanceProfileCredentialsProvider.java [hunk 1]
```diff
@@ -247,6 +250,25 @@ private URI getTokenEndpoint(String imdsHostname) {
         return URI.create(finalHost + TOKEN_RESOURCE);
     }
 
+    private String handleTokenErrorResponse(Exception e) {
+        if (isInsecureFallbackDisabled()) {
+            String message = String.format("Failed to retrieve IMDS token, and fallback to IMDS v1 is disabled via the %s "
+                                           + "environment variable and/or %s system property",
```
Comment (zoewangg) [inline core/auth/src/main/java/software/amazon/awssdk/auth/credentials/InstanceProfileCredentialsProvider.java:256]: and/or config as well?

File: core/auth/src/test/java/software/amazon/awssdk/auth/credentials/InstanceProfileCredentialsProviderTest.java [hunk 2]
```diff
@@ -45,11 +47,12 @@
 import java.util.List;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
-import org.junit.AfterClass;
-import org.junit.Before;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.rules.ExpectedException;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.BeforeEach;
```
Comment (zoewangg) [inline core/auth/src/test/java/software/amazon/awssdk/auth/credentials/InstanceProfileCredentialsProviderTest.java:51]: Nice, thanks for migrating to Junit5

File: core/auth/src/test/java/software/amazon/awssdk/auth/credentials/internal/Ec2MetadataDisableV1ResolverTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import software.amazon.awssdk.core.SdkSystemSetting;
+import software.amazon.awssdk.profiles.ProfileFile;
+import software.amazon.awssdk.profiles.ProfileProperty;
+import software.amazon.awssdk.testutils.EnvironmentVariableHelper;
+import software.amazon.awssdk.utils.Pair;
+import software.amazon.awssdk.utils.StringInputStream;
+
+@WireMockTest
```
Comment (zoewangg) [inline core/auth/src/test/java/software/amazon/awssdk/auth/credentials/internal/Ec2MetadataDisableV1ResolverTest.java:37]: nit: don't think we need wiremock in this test.

File: core/sdk-core/src/main/java/software/amazon/awssdk/core/SdkSystemSetting.java [hunk 4]
```diff
@@ -70,9 +70,14 @@ public enum SdkSystemSetting implements SystemSetting {
...
 
+    /**
+     * Whether to disable fallback to insecure EC2 Metadata instance service v1 on errors or timeouts.
+     */
+    AWS_EC2_METADATA_V1_DISABLED("aws.disableEc2MetadataV1", null),
+
     /**
      * The EC2 instance metadata service endpoint.
-     *
+     // *
```
Comment (zoewangg) [inline core/sdk-core/src/main/java/software/amazon/awssdk/core/SdkSystemSetting.java:80]: nit: unintentional change?

**Review by zoewangg — COMMENTED** [review]
I think we need to update EC2MetadataUtils to honor it as well; it's used in InstanceProfileRegionProvider. Can you do a search on "AWS_EC2_METADATA_DISABLED"? We need to make sure this new env is honored wherever AWS_EC2_METADATA_DISABLED is used IMO.

File: core/auth/src/test/java/software/amazon/awssdk/auth/credentials/internal/Ec2MetadataDisableV1ResolverTest.java (hunk 3 above)
Comment (cenedhryn) [reply core/auth/src/test/java/software/amazon/awssdk/auth/credentials/internal/Ec2MetadataDisableV1ResolverTest.java:37]: Copy paste... thanks

**Comment by cenedhryn** [comment]
> I think we need to update EC2MetadataUtils to honor it as well; it's used in InstanceProfileRegionProvider. Can you do a search on `AWS_EC2_METADATA_DISABLED`? We need to make sure this new env is honored wherever `AWS_EC2_METADATA_DISABLED` is used IMO.

Agreed offline that we do need to support the region provider at this point in time. Changed PR to push this code to a branch and will do a fast follow for region update. Once both PRs are approved, we can release the whole feature.

**Review by zoewangg — APPROVED** [review]

**Pull request merged**